usl_pipeline/usl_lib/usl_lib/transformers/elevation_transformers.py

```diff
@@ -0,0 +1,61 @@
+import typing
+from typing import Optional, Tuple
+
+import rasterio.features
+from shapely import geometry
+
+from usl_lib.readers import elevation_readers
+from usl_lib.transformers import polygon_transformers
+from usl_lib.writers import elevation_writers
+
+
+def transform_geotiff_with_boundaries_to_esri_ascii(
+    input_file: typing.BinaryIO,
+    output_file: typing.TextIO,
+    band: int = 1,
+    no_data_value: Optional[float] = None,
+    boundaries_polygons: Optional[list[Tuple[geometry.Polygon, int]]] = None,
+) -> None:
+    """Transforms elevation raster data from GeoTIFF file stream to Esri ASCII one.
+
+    Transformation optionally includes a stage of zeroing-out values (setting No-data)
+    for cells outside the polygon boundaries.
+
+    Args:
+        input_file: Binary stream to load from.
+        output_file: Textual stream to write to.
+        band: Index of a band that should be loaded from GeoTIFF.
+        no_data_value: Optional value to set in the returned data to indicate absence of
+            data. If not supplied, the no-data value defined in the TIFF file itself
+            will be used.
+        boundaries_polygons: Optional polygon boundaries that if defined would be used
+            to set No-data for elevation raster outside of boundaries.
+    """
+    with rasterio.open(input_file, driver="GTiff") as src:
+        input_nodata = src.nodata
+        header = elevation_readers.read_header_from_rasterio_dataset_reader(
+            src, no_data_value
+        )
+        elevation_writers.write_header_to_esri_ascii_raster_file(header, output_file)
+
+        if boundaries_polygons is not None:
+            boundaries_raster = polygon_transformers.rasterize_polygons(
+                header, boundaries_polygons
+            )
+
+        data = src.read(band)
+        row_index = 0
+        for row in data:
+            if no_data_value is not None:
+                row[row == input_nodata] = no_data_value
+            row_values = row.tolist()
+            # Zero-out elevation cells where boundaries raster is 0 (meaning outside the
+            # boundaries)
+            if boundaries_raster is not None:
+                boundaries_raster_row_values = boundaries_raster[row_index].tolist()
+                for col_index in range(len(row_values)):
```

Comment:
This loop looks equivalent to
```python
row[boundaries_raster[row_index] == 0] = no_data_value
```
or on the full dataset
```python
data[boundaries_raster == 0] = no_data_value
```
which will be actual orders of magnitude more performant than the loops. Vector operations in numpy are much much more performant than Python loops. Looping over a numpy array is something to always try to avoid.

Reply:
some numbers on a simple example
```python
import numpy
import timeit


def loops(arr):
    for i, row in enumerate(arr):
        for j, x in enumerate(row):
            arr[i, j] = x + 1


def vectors(arr):
    return arr + 1


for shape in ((500, 500), (5000, 5000), (10000, 10000), (15000, 15000)):
    arr = numpy.random.rand(*shape)
    number = 3
    print(
        f"loops {shape}",
        timeit.timeit(lambda: loops(arr), setup="pass", number=number) / number,
    )
    print(
        f"vectors {shape}",
        timeit.timeit(lambda: vectors(arr), setup="pass", number=number) / number,
    )
```
```
loops (500, 500) 0.026520140003412962
vectors (500, 500) 0.0004385486730219175
loops (5000, 5000) 2.7367708566598594
vectors (5000, 5000) 0.056308782659471035
loops (10000, 10000) 11.010823343327502
vectors (10000, 10000) 0.29148437933569465
loops (15000, 15000) 26.96899208734006
vectors (15000, 15000) 0.7613911676647452
```

Reply:
Yep, I knew about numpy being more efficient. Just realized that rasterio reads the whole thing into memory as numpy. Rewrote that to what you proposed, but with chunking.